torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+          grouped_inputs_outputs[1],
+          largest2,
+          0,
+          vectorize_factor2);
+
+  expanded_vector_word_size1 = std::min(
+      static_cast<size_t>(max_unroll_factor), expanded_vector_word_size1);
+  expanded_vector_word_size2 = std::min(
+      static_cast<size_t>(max_unroll_factor), expanded_vector_word_size2);
+
+  vectorize_factor1 = std::max(vectorize_factor1, expanded_vector_word_size1);
+  vectorize_factor2 = std::max(vectorize_factor2, expanded_vector_word_size2);
+
+  params.vectorize_factor1 = vectorize_factor1;
+  params.vectorize_factor2 = vectorize_factor2;
+
+  // TODO: should we adjust tile size according to max_unroll_factor?
+  // I have the following in mind, but seems not working
+  // size_t max_tile_elements =
+  //     std::min<size_t>(max_unroll_factor * kThreadsPerBlock, 32 * 32);
+  // size_t max_tile_size2 = static_cast<size_t>(std::sqrt(max_tile_elements));
+  // // tile_size2 must be a multiple of vectorize_factor2
+  // params.tile_size2 = max_tile_size2 / vectorize_factor2 * vectorize_factor2;
+  // size_t max_tile_size1 = max_tile_elements / params.tile_size2;
+  // // tile_size2 must be a multiple of vectorize_factor1
+  // params.tile_size1 = max_tile_size1 / vectorize_factor1 * vectorize_factor1;
+
+  auto warp_size = at::cuda::getCurrentDeviceProperties()->warpSize;
+  params.lparams.bind(warp_size, ParallelType::TIDx);
+  params.lparams.bind(kThreadsPerBlock / warp_size, ParallelType::TIDy);
+
+  if (isDebugDumpEnabled(DebugDumpOption::SchedulerDebug)) {
+    std::cerr << "\n===== Transpose Stats ========\n"
+              << "num_elems: " << n_elems << "\n"
+              << "n_input_tensors: " << n_input_tensors << "\n"
+              << "max_input_dtype_size: " << max_input_dtype_size << "\n"
+              << "group 1: " << ir_utils::toString(grouped_inputs_outputs[0])
+              << "\n"
+              << "group 2: " << ir_utils::toString(grouped_inputs_outputs[1])
+              << std::endl;
+    std::cerr << std::endl;
+    std::cerr << params.toString() << std::endl;
+  }
+
+  return params;
+}
+
+// TODO: remove or return launch parameters
+LaunchParams scheduleTranspose(
+    Fusion* fusion,
+    const at::ArrayRef<c10::IValue>& runtime_inputs) {
+  FUSER_PERF_SCOPE("scheduleFusion");
+  auto params = getTransposeHeuristics(fusion, runtime_inputs);
+  TORCH_INTERNAL_ASSERT(
+      params.has_value(), "Could not schedule pointwise operation.");
+  scheduleTranspose(fusion, params.value());
+  return params.value().lparams;
+}
+
+void scheduleTranspose(Fusion* fusion, const TransposeParams& params) {
+  FusionGuard fg(fusion);
+
+  // Make sure we don't have global memory set on intermediate tensors from
+  // fusion segmentation
+  scheduler_utils::clearMemorySpace(fusion);
+
+  // maybe has_reduction for scheduling should be done on a per output tensor
+  // basis.
+  TORCH_INTERNAL_ASSERT(
+      ir_utils::getReductionOps(fusion /*, ignore_trivial=true */).empty(),
```

Comment:
We should handle trivial reductions, but for this PR you should make sure this check in `canSchedule` matches.

Reply:
See https://github.com/csarofeen/pytorch/pull/1854#discussion_r931291189